### PR TITLE
Update jupyter packages to python 3 only packages with 27 support

### PR DESCRIPTION
**Pull request by costrouc**
###### Motivation for this change & Things done

Jupyter has moved their releases to [only supporting python 3+](https://github.com/jupyter/roadmap/blob/master/accepted/migration-to-python-3-only.md). This has blocked upgrading of several python packages within `nixpkgs`. These include `ipython`, `ipykernel`, and `jupyter`. This PR upgrades to newest releases while keeping python 2.7 support for these packages as well.

I would like input on if this is the appropriate way to support multiple versions. The important details are in the changes to `python-packages.nix`

This expression allows support for the most recent versions of `ipython` for each python version.

This update required making `prompt_toolkit` 2.x.x the default version. I have ensure that all applications that depended on prompt_toolkit 1.x.x still work. This issue was addressed recently with https://github.com/NixOS/nixpkgs/pull/49853.

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -2498,16 +2503,19 @@ in {
...
 
-  # Newer versions of IPython no longer support Python 2.7.
-  ipython = if isPy27 then self.ipython_5 else self.ipython_6;
-
-  ipython_5 = callPackage ../development/python-modules/ipython/5.nix { };
-
-  ipython_6 = callPackage ../development/python-modules/ipython { };
+  ipython = if isPy27 then
+      callPackage ../development/python-modules/ipython/5.nix { }
+    else if pythonOlder "3.5" then
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Don't bother with 2.7 < python < 3.5. Related: https://github.com/NixOS/nixpkgs/pull/49730

File: pkgs/development/python-modules/ipykernel/4.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+buildPythonPackage rec {
+  pname = "ipykernel";
+  version = "4.10.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "699103c8e64886e3ec7053f2a6aa83bb90426063526f63a818732ff385202bad";
+  };
+
+  buildInputs = [ nose ] ++ lib.optional isPy27 mock;
```
Comment (FRidh) [inline pkgs/development/python-modules/ipykernel/4.nix]: checkInputs

File: pkgs/development/python-modules/ipykernel/4.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  # Tests require backends.
+  # I don't want to add all supported backends as propagatedBuildInputs
+  doCheck = false;
+
+  meta = {
+    description = "IPython Kernel for Jupyter";
+    homepage = http://ipython.org/;
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ fridh ];
```
Comment (FRidh) [inline pkgs/development/python-modules/ipykernel/4.nix]: Can you remove me. I don't care about 2.7.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -3740,6 +3748,10 @@ in {
 
   prompt_toolkit = callPackage ../development/python-modules/prompt_toolkit { };
 
+  # incompatible change from 1.x.x -> 2.x.x
+  # required for jupyter packages (python 2.7 LTS)
+  prompt_toolkit_1 = callPackage ../development/python-modules/prompt_toolkit/1.nix { };
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Maybe we should always use this prompt toolkit with 2.7. That way we prevent possible collisions. I doubt packages require a newer version, but I may be wrong. We can keep it for a separate changeset.

**Comment by costrouc** [comment]
Bump. I should also add this this PR will enable "auto-install" of extensions which should make https://github.com/NixOS/nixpkgs/pull/49807 much easier.

**Comment by costrouc** [comment]
@GrahamcOfBorg build python36Packages.ipython python36Packages.prompt_toolkit python36Packages.notebook python36Packages.prompt_toolkit python36Packages.ipykernel python36Packages.jupyterlab_server python36Packages.jupyterlab python36Packages.jupyterhub

**Comment by Mic92** [comment]
@GrahamcOfBorg build python27Packages.jupyter python27Packages.jupyter_console

**Comment by Mic92** [comment]
https://github.com/Mic92/nix-review

**Comment by costrouc** [comment]
THANK YOU and written by none other than yourself. I guess I had seen that tool but never thought it was applicable to my use case (should have RTFM). I will do this to confirm my PRs in the future!

**Comment by costrouc** [comment]
I will also be closing this along with two other PRs to combine with this one. https://github.com/NixOS/nixpkgs/pull/49807,  https://github.com/NixOS/nixpkgs/pull/49721, and https://github.com/NixOS/nixpkgs/pull/50858.

This will lead to a PR with better python 3 support for jupyter (newest releases are python 3.5+ only). Addition of 3 kernels (c, go, ansible) and 20 jupyterlab extensions. Also will have an editor that has a clean way to add kernels and extensions. This will be a big PR but I am finding that they all depend on each other.

**Comment by FRidh** [comment]
I seem to have forgotten to mention it, but these changes have landed in master.

**Comment by FRidh** [comment]
An important change I made is that with Python 2 one always gets prompt_toolkit version 1, and with Python 3 always version 2.
